### DISCUSSION TITLE
Issues downloading files from a Codespace

### DISCUSSION THREAD

**Discussion by alfredodeza** [community/community#36166] [Codespaces]
Students doing ML work are facing issues trying to download files from a Codespace and getting an error message similar to this:

![Screen Shot 2022-10-14 at 12 09 18 PM](https://user-images.githubusercontent.com/317847/195903481-00df7eb7-4e2a-4e7a-90ab-aca3fc03b088.png)

This is problematic because it isn't clear what these "system files" may be even though the target directory is the _Downloads_ one. Clicking on the "Choose a different folder" gives another message that is not clear on the implications of accepting: 

![Screen Shot 2022-10-14 at 11 49 50 AM](https://user-images.githubusercontent.com/317847/195907443-5ef65228-ab88-4cd8-9cdf-8b9ad5dbac45.png)

In this case, is it true that the Codespace instance is going to edit files in the destination directory? 

Lastly, none of these messages appear with other _types_ of files. It only happens with a specific one: a downloaded model from HuggingFace with a `.bin` extension. 

## Reproduce

To replicate the issue in any Codespace instance:

1. Install the Transformers Huggingface library with `pip install transformers`
2. Download a model with:
```python
from transformers import pipeline

model = pipeline(
    "summarization",
    model="sshleifer/distilbart-cnn-12-6",
    revision="a4f8f3e",
)
model.save_pretrained("summarizeApp")
```
3. Attempt to download the resulting `pytorch_model.bin`

## Workarounds

Currently, this means granting access and accepting that Codespaces can edit files or create a new, empty directory with no contents that appears to not have any warning messages.

## Other related information

This issue only happens with Codespaces on the web. Issue is reproducible with Microsoft Edge and Google Chrome. It *does not* happen with Visual Studio Code running locally but connected to a remote Codespace.

Possibly unrelated: Download speeds seem to be throttled to less than 10mb/s

![Screen Shot 2022-10-14 at 11 50 30 AM](https://user-images.githubusercontent.com/317847/195908831-2635651f-be74-4edf-acf9-92a6b20fab83.png)

**Comment by joshspicer** [comment]
👋 , I believe this is a security feature of your browser's file system API.  In order for the file system API to let a website view/write to a directory on your computer, the destination folder needs to not contain files your browser deems "system files". I believe a file ending in `.bin` would qualify. 

I found similar discussion on the VS Code repo: https://github.com/microsoft/vscode/issues/144753#issuecomment-1063606119

Some workarounds i'd like to suggest are:

1. Use the [gh cs cp](https://cli.github.com/manual/gh_codespace_cp) command in the GitHub CLI to copy files.
2. Use the VS Code client to connect and download from there. 
3. Select a destination directory that doesn't have files that appear to be "system files" to avoid issues your browser's security features.  That may mean not using the web to download files that end in `.bin`
4. Attempt with a non-chromium browser (Safari or Firefox) and see if that may help. 
5. Start a quick python webserver (`python -m http.server`) and forward the port. You can then use port forwarding to download the file.

**Comment by alfredodeza** [comment]
Thanks for replying @joshspicer . I did go through the VSCode issue and read through the reference code. The issue I'm reporting is slightly different however. The destination folder *does not contain any system files*. Are you saying that downloading a ".bin" file from Codespaces to a folder would be the same as "destination containing system files" ?

As I covered in the report as well, even if that is the case, an empty `~/tmp` directory does work to download the .bin file. If I can't really download a `.bin` file, how is this allowed in an empty directory?
